tests/test_flow/test_persistence_random.py

```diff
@@ -87,57 +94,105 @@ def add_binding(self, out_descriptor, dep_entity_names, use_tuples_for_output):
 
         self._add_binding_to_flow(binding)
 
-    def update_binding(self, entity_name):
+    def update_binding(
+        self,
+        entity_name,
+        change_func_version=True,
+        update_version_annotation=True,
+    ):
         entity = self._entities_by_name[entity_name]
-        entity.binding.update_func_version()
+        entity.binding.update_func_version(
+            change_func=change_func_version,
+            update_annotation=update_version_annotation,
+            versioning_mode=self._versioning_mode,
+        )
 
         self._add_binding_to_flow(entity.binding)
 
     def query_and_check_entity(self, entity_name):
         flow = self._builder.build()
-        flow_value = flow.get(entity_name)
 
-        model_value = self._compute_model_value(entity_name)
+        flow_exception = None
+        try:
+            flow_value = flow.get(entity_name)
+        except CodeVersioningError as e:
+            flow_exception = e
+
+        model_exception = None
+        try:
+            model_value = self._compute_model_value(entity_name)
+        except CodeVersioningError as e:
+            model_exception = e
+
+        assert (flow_exception is None) == (model_exception is None)
+
+        if model_exception is None:
+            assert flow_value == model_value
+
+        else:
+            assert flow_exception.__class__ == model_exception.__class__
```

Comment:
Can you assert that the bad descriptors are also equal?

Reply:
I think it's possible for there to be multiple possible bad descriptors, and which one we get will depend on the order in which we traverse the dependencies. I can add a comment making this clear.

(In theory, after we make all the fixes we could assert that the two sets of bad descriptors are the same, but I'm not sure it's worth reorganizing the code to make this possible.)